<a href="https://colab.research.google.com/github/SantiagoBarraganPaipilla/SantiagoBarraganPaipilla/blob/main/Preparaci%C3%B3n_de_datos_en_R_PUJ.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Preparación de datos en R
## Probabilidad y estadística inferencial - Pontificia Universidad Javeriana
### Santiago Barragan Paipilla <br> Facultad de ingeniería

# Variables estadísticas

Una variable estadística es cualquier característica de interés para la investigación cuyo valor puede cambiar de un objeto a otro en la población.

De acuerdo con la medición que realizan, las variables se clasifican en cuantitativas y cualitativas.

1. **Variables cuantitativas:** También conocidas como variables *escalares*, las variables cuantitativas son las que toman como argumento cantidades numéricas (poseen unidades de medida). Son variables matemáticas.
    + Variables discretas: Presentan separaciones o interrupciones en la escala de valores que pueden tomar. Dichas separaciones indican la ausencia de valores que la variable puede asumir.
    * *Ejemplos:* número de hijos, nota de un parcial, dinero.
    + Variables continuas: Pueden tomar cualquier valor dentro de un intervalo o una unión de intervalos de números reales. Solamente están limitadas por la precisión del aparato (o unidad) medidor.
    * *Ejemplos:* temperatura, distancia, tiempo.

2. **Variables cualitativas** También llamadas categóricas. Las variables cualitativas son las que expresan cualidades, características o modalidad. Cada modalidad se llama atributo o categoría y la medición consiste en clasificar el objeto en su correspondiente categoría.
Se llaman dicotómicas cuando solo pueden tomar dos posibles valores y politómicas cuando tienen 3 o más categorías.

  Nota: Las categorías son de tipo clasificatorio y *no es posible* realizar operaciones entre ellas, ni siquiera cuando los atributos son codificados para facilitar la manipulación de los datos.
    * Variables nominales: Los atributos de la variable NO pueden ser sometidos a un criterio de orden objetivo y estándar.
    * *Ejemplos:* color del pelo, medio de transporte, carrera que se estudia, género, estado civil.
    * Variables ordinales: Las categorías de la variable pueden ordenarse de acuerdo con una escala objetiva.
    * *Ejemplos:* estrato socioeconómico, nivel académico, clasificación de triage en urgencias.

**tidyverse** es el paquete de R más utilizado y más eficiente para manipulación de datos. Vamos a instalarlo y a cargarlo en la sesión.

In [ ]:
install.packages("tidyverse")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘textshaping’, ‘conflicted’, ‘jsonlite’, ‘ragg’, ‘stringr’


Warning message in install.packages("tidyverse"):
“installation of package ‘textshaping’ had non-zero exit status”
Warning message in install.packages("tidyverse"):
“installation of package ‘ragg’ had non-zero exit status”
Warning message in install.packages("tidyverse"):
“installation of package ‘tidyverse’ had non-zero exit status”


In [ ]:
library(tidyverse)

Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.4.1     ✔ purrr   1.0.1
✔ tibble  3.2.1     ✔ dplyr   1.1.1
✔ tidyr   1.3.0     ✔ stringr 1.5.0
✔ readr   2.1.4     ✔ forcats 1.0.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



#### Lectura de datos

In [ ]:
# https://drive.google.com/file/d/1WzOZNbNjUpdwWHDYD4ExAsJC5tCjS8C_/view?usp=sharing
system("gdown --id 1WzOZNbNjUpdwWHDYD4ExAsJC5tCjS8C_")

In [ ]:
COVID_19 <- read.csv("COVID19_Colombia_Clase.csv", sep = ";")

A continuación, vamos a explorar cómo está leyendo R las columnas del conjunto de datos. Usamos la función glimpse de tidyverse.

In [ ]:
COVID_19 %>% glimpse()

Rows: 1,000,000
Columns: 13
$ FECHA_DE_INICIO_DE_SINTOMAS <chr> NA, "2021-05-26", "2020-12-27", "2020-12-2…
$ FECHA_DIAGNOSTICO           <chr> "2021-04-08", "2021-05-29", "2021-01-06", …
$ MES                         <chr> "abril", "mayo", "enero", "diciembre", "ju…
$ ANO                         <int> 2021, 2021, 2021, 2020, 2021, 2021, 2021, …
$ DIAS                        <int> NA, 3, 10, 5, 2, 2, 8, 5, 1, 6, 7, 2, NA, …
$ CIUDAD                      <chr> "Sin dato", "Bogota", "Bogota", "Bogota", …
$ LOCALIDAD_ASIS              <chr> "Sin dato", "Usaquen", "Suba", "Los Martir…
$ EDAD                        <int> 26, 33, 15, 41, 17, 34, 57, 1, 22, 6, 25, …
$ SEXO                        <chr> "F", "M", "M", "F", "M", "F", "M", "F", "M…
$ FUENTE_O_TIPO_DE_CONTAGIO   <chr> "En estudio", "En estudio", "Relacionado",…
$ UBICACION                   <chr> "Casa", "Casa", "Casa", "Casa", "Casa", "C…
$ ESTADO                      <chr> "Recuperado", "Recuperado", "Recuperado", …
$ SINTOMAS  

Note que no ejecutamos la función en la forma usual: **glimpse(COVID_19)** sino que usamos el operador *Pipe*: **COVID19 %>% glimpse()**.

El resultado es el mismo, pero la notación con **%>%** es más eficiente a la hora de manipular datos en R.

Analizando la salida de glimpse tenemos que las variables *AÑO*, *DIAS*, y *EDAD* están siendo interpretadas por R como numéricas (*int*) mientras que todas las demás son interpretadas como palabras (*chr*).

En la base, las variables *MES*, *CIUDAD*, *LOCALIDAD_ASIS*, *SEXO*, *FUENTE_O_TIPO_DE_CONTAGIO*, *UBICACION*, *ESTADO* y *SINTOMAS* son categóricas/cualitativas y, por tanto, R debe interpretarlas como factores y no como caracteres. Así, que debemos modificar su tipología.

Además, las variables *FECHA_INICIO_DE_SINTOMAS* y *FECHA_DIAGNOSTICO* al ser fechas no serán empleadas en nuestro posterior análisis descriptivo y, por tanto, vamos a retirarlas de la base de datos.

Ambos procedimientos los hacemos simultáneamente con las funciones *select* y *mutate* de **tidyverse**. (Note aquí la utilidad del operador %>% para encadenar instrucciones).

In [ ]:
COVID <- COVID_19 %>%
  select(-c(1,2)) %>%
    mutate_if(is.character, as.factor)

Inspeccionemos de nuevo la tipología de los datos

In [ ]:
COVID %>% glimpse()

Rows: 1,000,000
Columns: 11
$ MES                       <fct> abril, mayo, enero, diciembre, junio, mayo, …
$ ANO                       <int> 2021, 2021, 2021, 2020, 2021, 2021, 2021, 20…
$ DIAS                      <int> NA, 3, 10, 5, 2, 2, 8, 5, 1, 6, 7, 2, NA, 13…
$ CIUDAD                    <fct> Sin dato, Bogota, Bogota, Bogota, Bogota, Bo…
$ LOCALIDAD_ASIS            <fct> Sin dato, Usaquen, Suba, Los Martires, Teusa…
$ EDAD                      <int> 26, 33, 15, 41, 17, 34, 57, 1, 22, 6, 25, 14…
$ SEXO                      <fct> F, M, M, F, M, F, M, F, M, M, M, F, F, M, F,…
$ FUENTE_O_TIPO_DE_CONTAGIO <fct> En estudio, En estudio, Relacionado, En estu…
$ UBICACION                 <fct> Casa, Casa, Casa, Casa, Casa, Casa, Casa, Ca…
$ ESTADO                    <fct> Recuperado, Recuperado, Recuperado, Recupera…
$ SINTOMAS                  <fct> NO, SI, SI, SI, SI, SI, SI, SI, SI, SI, SI, …


¡Perfecto!

### Conteos

Con las funciones *distinct* y *count* podemos establecer, respectivamente, cuáles son los niveles/categorías de una variable cualitativa y cuántos registros hay en cada uno. Veamos:

In [ ]:
COVID %>% distinct(UBICACION)

UBICACION            
1 Casa                 
2 Fallecido            
3 Fallecido (No aplica)
4 Hospital             
5 Hospital UCI

In [ ]:
COVID %>% count(UBICACION)

UBICACION             n     
1 Casa                  977656
2 Fallecido              19103
3 Fallecido (No aplica)   2965
4 Hospital                 135
5 Hospital UCI             141

Es decir, en la base de datos hay 135 personas reportadas en el hospital al momento de hacer la medición y 19103 que aparecen registradas como fallecidas.

### Filtros

Vamos ahora a filtrar la base de datos en dos bases separadas: una para las personas que presentaron síntomas y otra para los pacientes asintomáticos.

Para ello usaremos la función *filter* de **tidyverse** y la variable *SINTOMAS* de la base.

In [ ]:
COVID_ASIN <- COVID %>% filter(SINTOMAS == "NO")
COVID_CON <- COVID %>% filter(SINTOMAS == "SI")